Src/crypto/chacha20.c

```diff
@@ -0,0 +1,175 @@
+        0U, 0U, 0U, 0U, 0U, 0U, 0U, 0U, \
+        0U, 0U, 0U, 0U, 0U, 0U, 0U, 0U, \
+        0U, 0U, 0U, 0U, 0U, 0U, 0U, 0U  \
+    }
+
+static inline void quarterround(uint32_t* a, uint32_t* b, uint32_t* c, uint32_t* d);
+static void chacha20_block(const byte_t key[CHACHA20_KEY_SIZE], const uint32_t counter,
+                           const byte_t nonce[CHACHA20_NONCE_SIZE],
+                           byte_t out[KEY_STREAM_SIZE]);
+
+static inline void
+quarterround(uint32_t* a, uint32_t* b, uint32_t* c, uint32_t* d) {
+    *a += *b;
+    *d = BitManipulation_rotl32(*d ^ *a, 16);
+    *c += *d;
+    *b = BitManipulation_rotl32(*b ^ *c, 12);
+    *a += *b;
+    *d = BitManipulation_rotl32(*d ^ *a, 8);
+    *c += *d;
+    *b = BitManipulation_rotl32(*b ^ *c, 7);
+}
+
+static void
+chacha20_block(const byte_t key[CHACHA20_KEY_SIZE], const uint32_t counter, const byte_t nonce[CHACHA20_NONCE_SIZE],
+               byte_t out[KEY_STREAM_SIZE]) {
+
+    uint32_t x0  = CHACHA_MAGIC_NUMBER_PART_1;
+    uint32_t x1  = CHACHA_MAGIC_NUMBER_PART_2;
+    uint32_t x2  = CHACHA_MAGIC_NUMBER_PART_3;
+    uint32_t x3  = CHACHA_MAGIC_NUMBER_PART_4;
+    uint32_t x4  = Utils_Deserialize32LE(&key[0]);
+    uint32_t x5  = Utils_Deserialize32LE(&key[4]);
+    uint32_t x6  = Utils_Deserialize32LE(&key[8]);
+    uint32_t x7  = Utils_Deserialize32LE(&key[12]);
+    uint32_t x8  = Utils_Deserialize32LE(&key[16]);
+    uint32_t x9  = Utils_Deserialize32LE(&key[20]);
+    uint32_t x10 = Utils_Deserialize32LE(&key[24]);
+    uint32_t x11 = Utils_Deserialize32LE(&key[28]);
+    uint32_t x12 = counter;
+    uint32_t x13 = Utils_Deserialize32LE(&nonce[0]);
+    uint32_t x14 = Utils_Deserialize32LE(&nonce[4]);
+    uint32_t x15 = Utils_Deserialize32LE(&nonce[8]);
+
+    for (uint8_t i = 0; i < 10U; ++i) {
+        quarterround(&x0, &x4,  &x8, &x12);
+        quarterround(&x1, &x5,  &x9, &x13);
+        quarterround(&x2, &x6, &x10, &x14);
+        quarterround(&x3, &x7, &x11, &x15);
+        quarterround(&x0, &x5, &x10, &x15);
+        quarterround(&x1, &x6, &x11, &x12);
+        quarterround(&x2, &x7,  &x8, &x13);
+        quarterround(&x3, &x4,  &x9, &x14);
+    }
+
+    x0  += CHACHA_MAGIC_NUMBER_PART_1;
+    x1  += CHACHA_MAGIC_NUMBER_PART_2;
+    x2  += CHACHA_MAGIC_NUMBER_PART_3;
+    x3  += CHACHA_MAGIC_NUMBER_PART_4;
+    x4  += Utils_Deserialize32LE(&key[0]);
+    x5  += Utils_Deserialize32LE(&key[4]);
+    x6  += Utils_Deserialize32LE(&key[8]);
+    x7  += Utils_Deserialize32LE(&key[12]);
+    x8  += Utils_Deserialize32LE(&key[16]);
+    x9  += Utils_Deserialize32LE(&key[20]);
+    x10 += Utils_Deserialize32LE(&key[24]);
+    x11 += Utils_Deserialize32LE(&key[28]);
+    x12 += counter;
+    x13 += Utils_Deserialize32LE(&nonce[0]);
+    x14 += Utils_Deserialize32LE(&nonce[4]);
+    x15 += Utils_Deserialize32LE(&nonce[8]);
+
+    /* -E> compliant MC3R1.R18.6 16 automatic storage pointed with "out" is not copied to values from this function */

```

Comment:
byte* result = out? And pass result in functions? Same as in Sort algorithms.

Reply:
it didn't change the outcome. I'll keep it as it is since I don't see how this violates R18.6 

![image](https://github.com/IMProject/IMUtility/assets/10188706/241bb8ff-e06d-4dbf-abce-04aae841d6ba)